src/Types.jl

```diff
@@ -459,20 +459,60 @@ end
 function write_env_usage(source_file::AbstractString, usage_filepath::AbstractString)
     # Don't record ghost usage
     !isfile(source_file) && return
-
     # Ensure that log dir exists
     !ispath(logdir()) && mkpath(logdir())
 
-    # Generate entire entry as a string first
-    entry = sprint() do io
-        TOML.print(io, Dict(source_file => [Dict("time" => now())]))
-    end
-
-    # Append entry to log file in one chunk
+    p = TOML.Parser()
     usage_file = joinpath(logdir(), usage_filepath)
-    open(usage_file, append=true) do io
-        write(io, entry)
+    timestamp = now()
+
+    ## Atomically write usage file
+    while true
+        # read existing usage file
+        usage = if isfile(usage_file)
+            content = read(usage_file)
```

Comment:
Isn't this block just `TOML.parse(p, usage_file)`?

Reply:
I also think it is fine to just use the `TOML.parse(usage_file` API here. In the vast majority of cases you will only read this file twice so it isn't very useful to save the parser object.